File: android/app/src/test/kotlin/is/borgarland/poc/TelemetryTest.kt
package `is`.borgarland.poc

import `is`.borgarland.poc.net.Telemetry
import `is`.borgarland.poc.net.TelemetryEvent
import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The client half of the telemetry channel, pinned to data/relay-events.json.
 * The contract is the privacy boundary, so these tests are about what can and
 * cannot reach the wire: the envelope shape, the session format, the batch
 * cap, and above all that a description string handed anywhere near the
 * channel never appears in an encoded body.
 */
class TelemetryTest {

    /** The common case: the relay takes every batch it is handed. */
    private fun telemetryWith(send: (String) -> Unit): Telemetry =
        Telemetry().apply {
            this.send = { body, done ->
                send(body)
                done(Telemetry.BatchOutcome.DELIVERED)
            }
        }

    /** The offline case: the caller decides what the relay did with each batch. */
    private fun telemetryWith(outcome: () -> Telemetry.BatchOutcome, send: (String) -> Unit): Telemetry =
        Telemetry().apply {
            this.send = { body, done ->
                send(body)
                done(outcome())
            }
        }

    private fun namesIn(body: String): List<String> =
        Json.parseToJsonElement(body).jsonObject.getValue("events").jsonArray
            .map { it.jsonObject.getValue("name").jsonPrimitive.content }

    @Test
    fun `envelope carries the contract shape`() {
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith { bodies += it }

        telemetry.track(TelemetryEvent.AppOpened)
        telemetry.track(TelemetryEvent.SendStarted)
        telemetry.flush()

        val json = Json.parseToJsonElement(bodies.single()).jsonObject
        assertTrue("session must be 32 lowercase hex", json.getValue("session").jsonPrimitive.content.matches(Regex("[0-9a-f]{32}")))
        assertEquals("android", json.getValue("platform").jsonPrimitive.content)
        assertEquals(
            "${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
            json.getValue("appVersion").jsonPrimitive.content,
        )

        val events = json.getValue("events").jsonArray
        assertEquals(2, events.size)
        assertEquals("app-opened", events[0].jsonObject.getValue("name").jsonPrimitive.content)
        assertEquals("send-started", events[1].jsonObject.getValue("name").jsonPrimitive.content)
        // atMs is the offset from session start, within a day.
        assertTrue(
            events[0].jsonObject.getValue("atMs").jsonPrimitive.content.toInt() in 0..86_400_000,
        )
        // app-opened carries no fields at all — exactly name + atMs.
        assertEquals(2, events[0].jsonObject.size)
    }

    @Test
    fun `field carrying event encodes its fields`() {
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith { bodies += it }

        telemetry.track(TelemetryEvent.LocationResolved(5, TelemetryEvent.LocationSource.DEVICE, 7))
        telemetry.flush()

        val event = Json.parseToJsonElement(bodies.single()).jsonObject
            .getValue("events").jsonArray[0].jsonObject
        assertEquals("location-resolved", event.getValue("name").jsonPrimitive.content)
        assertEquals(5, event.getValue("elapsedMs").jsonPrimitive.content.toInt())
        assertEquals("device", event.getValue("source").jsonPrimitive.content)
        assertEquals(7, event.getValue("accuracyM").jsonPrimitive.content.toInt())
    }

    @Test
    fun `session is fresh per instance`() {
        // One instance per app LAUNCH, so two instances must never share a
        // session: the id is what makes the stream a timeline rather than a
        // tracker.
        assertNotEquals(Telemetry().sessionId, Telemetry().sessionId)
    }

    @Test
    fun `flushes when the buffer reaches the threshold`() {
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith { bodies += it }

        repeat(19) { telemetry.track(TelemetryEvent.AppOpened) }
        assertTrue("the buffer must hold until the threshold", bodies.isEmpty())
        telemetry.track(TelemetryEvent.AppOpened)
        assertEquals("the 20th event must trigger a flush", 1, bodies.size)
    }

    @Test
    fun `batch never exceeds one hundred and drops the oldest`() {
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith { bodies += it }
        telemetry.flushThreshold = Int.MAX_VALUE

        // Distinct byte counts identify the order of the buffered events.
        for (i in 0 until 150) {
            telemetry.track(TelemetryEvent.PhotoCaptured(i, i, "image/jpeg"))
        }
        telemetry.flush()

        val events = Json.parseToJsonElement(bodies.single()).jsonObject.getValue("events").jsonArray
        assertEquals("the relay refuses a batch longer than 100", 100, events.size)
        assertEquals("the oldest 50 must have been dropped", 50, events.first().jsonObject.getValue("bytes").jsonPrimitive.content.toInt())
        assertEquals(149, events.last().jsonObject.getValue("bytes").jsonPrimitive.content.toInt())
    }

    @Test
    fun `consecutive description lengths coalesce`() {
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith { bodies += it }

        // One event per keystroke would spam the relay; the buffer keeps the
        // latest length only.
        telemetry.track(TelemetryEvent.DescriptionLength(1))
        telemetry.track(TelemetryEvent.DescriptionLength(5))
        telemetry.track(TelemetryEvent.DescriptionLength(9))
        telemetry.flush()

        val events = Json.parseToJsonElement(bodies.single()).jsonObject.getValue("events").jsonArray
        assertEquals(1, events.size)
        assertEquals(9, events.single().jsonObject.getValue("length").jsonPrimitive.content.toInt())
    }

    // A batch the relay did not take.

    /**
     * #74: the airplane-mode walk of the first iOS field test is missing from
     * D1 entirely. The buffer was emptied before the request, the request
     * failed, and a two-minute hole in the timeline is indistinguishable from
     * a tester standing still. The Kotlin has the same shape and the same bug.
     */
    @Test
    fun `an undelivered batch is kept and sent again`() {
        val bodies = mutableListOf<String>()
        var outcome = Telemetry.BatchOutcome.UNDELIVERED
        val telemetry = telemetryWith({ outcome }) { bodies += it }

        telemetry.track(TelemetryEvent.CategoryChosen(1, "ruslafotur"))
        telemetry.track(TelemetryEvent.SendStarted)
        telemetry.flush()
        assertEquals("the first attempt is made", 1, bodies.size)

        // The network came back, and nothing else happened in between.
        outcome = Telemetry.BatchOutcome.DELIVERED
        telemetry.flush()

        assertEquals("the failed batch is tried again", 2, bodies.size)
        assertEquals(
            "and it carries the events the failed attempt held",
            listOf("category-chosen", "send-started"),
            namesIn(bodies[1]),
        )
    }

    @Test
    fun `a requeued batch stays ahead of what came after it`() {
        val bodies = mutableListOf<String>()
        var outcome = Telemetry.BatchOutcome.UNDELIVERED
        val telemetry = telemetryWith({ outcome }) { bodies += it }

        telemetry.track(TelemetryEvent.SendStarted)
        telemetry.flush()

        // Buffered while the failed request was still out, so newer.
        telemetry.track(TelemetryEvent.ScreenLeft(TelemetryEvent.Screen.SUMMARY, false))
        outcome = Telemetry.BatchOutcome.DELIVERED
        telemetry.flush()

        assertEquals(
            "a batch put back is older than what arrived while it was in flight",
            listOf("send-started", "screen-left"),
            namesIn(bodies[1]),
        )
    }

    /**
     * The other half of the fix. A relay that refuses this body will refuse it
     * again, so a rejected batch must be dropped or it blocks the buffer for
     * the rest of the session.
     */
    @Test
    fun `a rejected batch is dropped rather than tried forever`() {
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith({ Telemetry.BatchOutcome.REJECTED }) { bodies += it }

        telemetry.track(TelemetryEvent.AppOpened)
        telemetry.flush()
        telemetry.flush()

        assertEquals("nothing was left to send", 1, bodies.size)
    }

    @Test
    fun `status is read as keep or drop`() {
        // The relay took it.
        assertEquals(Telemetry.BatchOutcome.DELIVERED, Telemetry.outcomeForStatus(200))
        assertEquals(Telemetry.BatchOutcome.DELIVERED, Telemetry.outcomeForStatus(204))
        // The relay refused this body and would refuse it again.
        assertEquals(Telemetry.BatchOutcome.REJECTED, Telemetry.outcomeForStatus(400))
        assertEquals(Telemetry.BatchOutcome.REJECTED, Telemetry.outcomeForStatus(413))
        // Try later: a timeout, a rate limit, or the relay being down.
        assertEquals(Telemetry.BatchOutcome.UNDELIVERED, Telemetry.outcomeForStatus(408))
        assertEquals(Telemetry.BatchOutcome.UNDELIVERED, Telemetry.outcomeForStatus(429))
        assertEquals(Telemetry.BatchOutcome.UNDELIVERED, Telemetry.outcomeForStatus(500))
        assertEquals(Telemetry.BatchOutcome.UNDELIVERED, Telemetry.outcomeForStatus(0))
    }

    /**
     * Without this, every event past the threshold starts another post while
     * offline, and a batch requeued underneath a later one duplicates it.
     */
    @Test
    fun `only one batch is in flight at a time`() {
        val bodies = mutableListOf<String>()
        val pending = mutableListOf<(Telemetry.BatchOutcome) -> Unit>()
        val telemetry = Telemetry().apply {
            this.send = { body, done ->
                bodies += body
                pending += done
            }
        }

        telemetry.track(TelemetryEvent.AppOpened)
        telemetry.flush()
        assertEquals(1, bodies.size)

        telemetry.track(TelemetryEvent.SendStarted)
        telemetry.flush()
        assertEquals("the second flush waits for the first to answer", 1, bodies.size)

        pending[0](Telemetry.BatchOutcome.DELIVERED)
        telemetry.flush()
        assertEquals(listOf("send-started"), namesIn(bodies[1]))
    }

    /**
     * Dropping under the cap stays correct: the requeued batch is not
     * privileged, and the oldest events still go when there are too many.
     */
    @Test
    fun `the cap still applies to a requeued batch`() {
        val bodies = mutableListOf<String>()
        var outcome = Telemetry.BatchOutcome.UNDELIVERED
        val telemetry = telemetryWith({ outcome }) { bodies += it }
        telemetry.flushThreshold = Int.MAX_VALUE

        // Distinct byte counts identify the order of the buffered events.
        for (i in 0 until 100) {
            telemetry.track(TelemetryEvent.PhotoCaptured(i, i, "image/jpeg"))
        }
        telemetry.flush()
        for (i in 100 until 150) {
            telemetry.track(TelemetryEvent.PhotoCaptured(i, i, "image/jpeg"))
        }
        outcome = Telemetry.BatchOutcome.DELIVERED
        telemetry.flush()

        val events = Json.parseToJsonElement(bodies[1]).jsonObject.getValue("events").jsonArray
        assertEquals(100, events.size)
        assertEquals("the oldest 50 went, requeued or not", 50, events.first().jsonObject.getValue("bytes").jsonPrimitive.content.toInt())
        assertEquals(149, events.last().jsonObject.getValue("bytes").jsonPrimitive.content.toInt())
    }

    @Test
    fun `description text never appears in the encoded body`() {
        val secret = "FULL RUSLAFATA VIÐ STÍGINN 83721"
        val bodies = mutableListOf<String>()
        val telemetry = telemetryWith { bodies += it }

        // The closest the app ever brings a description to this channel:
        // only its length is passed, never the text.
        telemetry.track(TelemetryEvent.DescriptionLength(secret.length))
        telemetry.track(TelemetryEvent.CategoryChosen(1, "ruslafotur"))
        telemetry.flush()

        val body = bodies.single()
        assertFalse("the description text must never reach the wire", body.contains(secret))
        // No description field exists anywhere in the contract; a stray one
        // would be rejected by the relay and is a privacy violation here.
        assertFalse(body.contains("\"description\""))
        // The length is the whole signal, and it is present.
        assertTrue(body.contains("\"length\":${secret.length}"))
    }

    @Test
    fun `event names match the contract exactly`() {
        // The unit tests run with the module directory as the working
        // directory, so the repo root is two levels up (FactsFileTest reads
        // src/main/assets the same way).
        val text = File("../../data/relay-events.json").readText()
        val contractEvents = Json.parseToJsonElement(text).jsonObject.getValue("events").jsonObject
        assertEquals(TelemetryEvent.allNames, contractEvents.keys)
    }

    @Test
    fun `event fields match the contract exactly`() {
        val text = File("../../data/relay-events.json").readText()
        val contractEvents = Json.parseToJsonElement(text).jsonObject.getValue("events").jsonObject

        for (name in TelemetryEvent.allNames) {
            val contractFields = contractEvents.getValue(name).jsonObject
                .getValue("fields").jsonObject.keys
            assertEquals(
                "field set for $name drifted from the contract",
                contractFields,
                sampleEvent(name).fields.keys,
            )
        }
    }

    /** One representative event per name, used to pin the field sets. A new
     * field on a case is a compile error here, which is the point: the drift
     * is caught at build time, not by the relay's 400. */
    private fun sampleEvent(name: String): TelemetryEvent = when (name) {
        "app-opened" -> TelemetryEvent.AppOpened
        "camera-permission" -> TelemetryEvent.CameraPermission(true)
        "location-permission" -> TelemetryEvent.LocationPermission(true)
        "photo-captured" -> TelemetryEvent.PhotoCaptured(0, 0, "image/jpeg")
        "location-resolved" -> TelemetryEvent.LocationResolved(0, TelemetryEvent.LocationSource.DEVICE, 0)
        "location-failed" -> TelemetryEvent.LocationFailed(0, TelemetryEvent.LocationFailure.TIMEOUT)
        "category-chosen" -> TelemetryEvent.CategoryChosen(0, "ruslafotur")
        "description-length" -> TelemetryEvent.DescriptionLength(0)
        "send-started" -> TelemetryEvent.SendStarted
        "send-result" -> TelemetryEvent.SendResult(0, 200, true)
        "send-failed" -> TelemetryEvent.SendFailed(0, TelemetryEvent.SendFailure.CONNECTION)
        "screen-left" -> TelemetryEvent.ScreenLeft(TelemetryEvent.Screen.CAMERA, true)
        else -> error("unknown event name $name")
    }
}

File: ios/BorgarlandCore/Tests/BorgarlandCoreTests/TelemetryTest.swift
import XCTest
@testable import BorgarlandCore

/// The client half of the telemetry channel, pinned to data/relay-events.json.
/// The contract is the privacy boundary, so these tests are about what can
/// and cannot reach the wire: the envelope shape, the session format, the
/// batch cap, and above all that a description string handed anywhere near
/// the channel never appears in an encoded body.
final class TelemetryTest: XCTestCase {

    // MARK: - Envelope

    func testEnvelopeCarriesTheContractShape() throws {
        let start = Date(timeIntervalSince1970: 1_000)
        var bodies: [Data] = []
        let telemetry = Telemetry(
            sessionID: String(repeating: "ab", count: 16),
            sessionStart: start,
            now: { start.addingTimeInterval(1.5) }
        )
        telemetry.appVersion = "0.1.0 (3)"
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }

        telemetry.track(.appOpened)
        telemetry.track(.sendStarted)
        telemetry.flush()

        let json = try XCTUnwrap(JSONSerialization.jsonObject(with: try XCTUnwrap(bodies.first)) as? [String: Any])
        XCTAssertEqual(json["session"] as? String, String(repeating: "ab", count: 16))
        XCTAssertEqual(json["platform"] as? String, "ios")
        XCTAssertEqual(json["appVersion"] as? String, "0.1.0 (3)")

        let events = try XCTUnwrap(json["events"] as? [[String: Any]])
        XCTAssertEqual(events.count, 2)
        // atMs is the offset from session start, recorded per event.
        XCTAssertEqual(events[0]["name"] as? String, "app-opened")
        XCTAssertEqual(events[0]["atMs"] as? Int, 1500)
        // app-opened carries no fields at all — exactly name + atMs.
        XCTAssertEqual(events[0].count, 2)
        XCTAssertEqual(events[1]["name"] as? String, "send-started")
    }

    func testFieldCarryingEventEncodesItsFields() throws {
        let start = Date()
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }

        telemetry.track(.locationResolved(elapsedMs: 5, source: .device, accuracyM: 7))
        telemetry.flush()

        let events = try XCTUnwrap(
            (JSONSerialization.jsonObject(with: try XCTUnwrap(bodies.first)) as? [String: Any])?["events"] as? [[String: Any]]
        )
        XCTAssertEqual(events[0]["name"] as? String, "location-resolved")
        XCTAssertEqual(events[0]["elapsedMs"] as? Int, 5)
        XCTAssertEqual(events[0]["source"] as? String, "device")
        XCTAssertEqual(events[0]["accuracyM"] as? Int, 7)
    }

    func testAtMsIsClampedToADay() throws {
        let start = Date(timeIntervalSince1970: 0)
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start.addingTimeInterval(2 * 86_400) })
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }

        telemetry.track(.appOpened)
        telemetry.flush()

        let events = try XCTUnwrap(
            (JSONSerialization.jsonObject(with: try XCTUnwrap(bodies.first)) as? [String: Any])?["events"] as? [[String: Any]]
        )
        XCTAssertEqual(events[0]["atMs"] as? Int, 86_400_000)
    }

    // MARK: - Session

    func testSessionIs32LowercaseHexCharacters() {
        let session = Telemetry.newSessionID()
        XCTAssertEqual(session.count, 32)
        XCTAssertNil(session.range(of: "[^0-9a-f]", options: .regularExpression))
    }

    func testSessionIsFreshPerInstance() {
        // One instance per app LAUNCH, so two instances must never share a
        // session: the id is what makes the stream a timeline rather than a
        // tracker.
        XCTAssertNotEqual(Telemetry().sessionID, Telemetry().sessionID)
    }

    // MARK: - Batching

    func testFlushesWhenTheBufferReachesTheThreshold() {
        let start = Date()
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }

        for _ in 0..<19 {
            telemetry.track(.appOpened)
        }
        XCTAssertTrue(bodies.isEmpty, "the buffer must hold until the threshold")
        telemetry.track(.appOpened)
        XCTAssertEqual(bodies.count, 1, "the 20th event must trigger a flush")
    }

    func testBatchNeverExceeds100AndDropsTheOldest() throws {
        let start = Date()
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start.addingTimeInterval(0.001) })
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }
        telemetry.flushThreshold = .max

        // Distinct byte counts identify the order of the buffered events.
        for i in 0..<150 {
            telemetry.track(.photoCaptured(elapsedMs: i, bytes: i, mime: "image/jpeg"))
        }
        telemetry.flush()

        let events = try XCTUnwrap(
            (JSONSerialization.jsonObject(with: try XCTUnwrap(bodies.first)) as? [String: Any])?["events"] as? [[String: Any]]
        )
        XCTAssertEqual(events.count, 100, "the relay refuses a batch longer than 100")
        XCTAssertEqual(events.first?["bytes"] as? Int, 50, "the oldest 50 must have been dropped")
        XCTAssertEqual(events.last?["bytes"] as? Int, 149)
    }

    func testConsecutiveDescriptionLengthsCoalesce() throws {
        let start = Date()
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }

        // One event per keystroke would spam the relay; the buffer keeps the
        // latest length only.
        telemetry.track(.descriptionLength(length: 1))
        telemetry.track(.descriptionLength(length: 5))
        telemetry.track(.descriptionLength(length: 9))
        telemetry.flush()

        let events = try XCTUnwrap(
            (JSONSerialization.jsonObject(with: try XCTUnwrap(bodies.first)) as? [String: Any])?["events"] as? [[String: Any]]
        )
        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0]["length"] as? Int, 9)
    }

    // MARK: - A batch the relay did not take

    /// #74, and the reason this section exists: the airplane-mode walk of the
    /// first iOS field test is missing from D1 entirely. The buffer was emptied
    /// before the request, the request failed, and a two-minute hole in the
    /// timeline is indistinguishable from a tester standing still.
    func testAnUndeliveredBatchIsKeptAndSentAgain() throws {
        let start = Date()
        var bodies: [Data] = []
        var outcome = Telemetry.BatchOutcome.undelivered
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(outcome) }

        telemetry.track(.categoryChosen(elapsedMs: 1, slug: "ruslafotur"))
        telemetry.track(.sendStarted)
        telemetry.flush()
        XCTAssertEqual(bodies.count, 1, "the first attempt is made")

        // The network came back, and nothing else happened in between.
        outcome = .delivered
        telemetry.flush()

        XCTAssertEqual(bodies.count, 2, "the failed batch is tried again")
        XCTAssertEqual(
            try names(in: bodies[1]),
            ["category-chosen", "send-started"],
            "and it carries the events the failed attempt held"
        )
    }

    func testARequeuedBatchStaysAheadOfWhatCameAfterIt() throws {
        let start = Date()
        var bodies: [Data] = []
        var outcome = Telemetry.BatchOutcome.undelivered
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(outcome) }

        telemetry.track(.sendStarted)
        telemetry.flush()

        // Buffered while the failed request was still out, so newer.
        telemetry.track(.screenLeft(screen: .confirm, completed: false))
        outcome = .delivered
        telemetry.flush()

        XCTAssertEqual(
            try names(in: bodies[1]),
            ["send-started", "screen-left"],
            "a batch put back is older than what arrived while it was in flight"
        )
    }

    /// The other half of the fix. A relay that refuses this body will refuse it
    /// again — the `noExif` bug 400'd every batch that contained one — so a
    /// rejected batch must be dropped or it blocks the buffer for the session.
    func testARejectedBatchIsDroppedRatherThanTriedForever() throws {
        let start = Date()
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(.rejected) }

        telemetry.track(.appOpened)
        telemetry.flush()
        telemetry.flush()

        XCTAssertEqual(bodies.count, 1, "nothing was left to send")
    }

    func testStatusIsReadAsKeepOrDrop() {
        // The relay took it.
        XCTAssertEqual(Telemetry.outcome(forStatus: 200), .delivered)
        XCTAssertEqual(Telemetry.outcome(forStatus: 204), .delivered)
        // The relay refused this body and would refuse it again.
        XCTAssertEqual(Telemetry.outcome(forStatus: 400), .rejected)
        XCTAssertEqual(Telemetry.outcome(forStatus: 413), .rejected)
        // Try later: a timeout, a rate limit, or the relay being down.
        XCTAssertEqual(Telemetry.outcome(forStatus: 408), .undelivered)
        XCTAssertEqual(Telemetry.outcome(forStatus: 429), .undelivered)
        XCTAssertEqual(Telemetry.outcome(forStatus: 500), .undelivered)
        XCTAssertEqual(Telemetry.outcome(forStatus: 0), .undelivered)
    }

    /// Without this, every event past the threshold starts another post while
    /// offline, and a batch requeued underneath a later one duplicates it.
    func testOnlyOneBatchIsInFlightAtATime() throws {
        let start = Date()
        var bodies: [Data] = []
        var pending: [(Telemetry.BatchOutcome) -> Void] = []
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); pending.append(done) }

        telemetry.track(.appOpened)
        telemetry.flush()
        XCTAssertEqual(bodies.count, 1)

        telemetry.track(.sendStarted)
        telemetry.flush()
        XCTAssertEqual(bodies.count, 1, "the second flush waits for the first to answer")

        pending[0](.delivered)
        telemetry.flush()
        XCTAssertEqual(try names(in: bodies[1]), ["send-started"])
    }

    /// Dropping under the cap stays correct: the requeued batch is not
    /// privileged, and the oldest events still go when there are too many.
    func testTheCapStillAppliesToARequeuedBatch() throws {
        let start = Date()
        var bodies: [Data] = []
        var outcome = Telemetry.BatchOutcome.undelivered
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(outcome) }
        telemetry.flushThreshold = .max

        // Distinct byte counts identify the order of the buffered events.
        for i in 0..<100 {
            telemetry.track(.photoCaptured(elapsedMs: i, bytes: i, mime: "image/jpeg"))
        }
        telemetry.flush()
        for i in 100..<150 {
            telemetry.track(.photoCaptured(elapsedMs: i, bytes: i, mime: "image/jpeg"))
        }
        outcome = .delivered
        telemetry.flush()

        let events = try decodedEvents(in: bodies[1])
        XCTAssertEqual(events.count, 100)
        XCTAssertEqual(events.first?["bytes"] as? Int, 50, "the oldest 50 went, requeued or not")
        XCTAssertEqual(events.last?["bytes"] as? Int, 149)
    }

    // MARK: - Reading a body back

    private func decodedEvents(in body: Data) throws -> [[String: Any]] {
        try XCTUnwrap(
            (JSONSerialization.jsonObject(with: body) as? [String: Any])?["events"] as? [[String: Any]]
        )
    }

    private func names(in body: Data) throws -> [String] {
        try decodedEvents(in: body).map { $0["name"] as? String ?? "" }
    }

    // MARK: - The privacy test

    func testDescriptionTextNeverAppearsInTheBody() throws {
        let secret = "FULL RUSLAFATA VIÐ STÍGINN 83721"
        let start = Date()
        var bodies: [Data] = []
        let telemetry = Telemetry(sessionStart: start, now: { start })
        telemetry.send = { body, done in bodies.append(body); done(.delivered) }

        // The closest the app ever brings a description to this channel:
        // only its length is passed, never the text.
        telemetry.track(.descriptionLength(length: secret.count))
        telemetry.track(.categoryChosen(elapsedMs: 1, slug: "ruslafotur"))
        telemetry.flush()

        // Two unwraps, not one: `bodies.first` is optional and the String
        // initialiser is failable, so `.map` produces a String?? and a single
        // XCTUnwrap leaves an optional behind.
        let data = try XCTUnwrap(bodies.first)
        let body = try XCTUnwrap(String(data: data, encoding: .utf8))
        XCTAssertFalse(body.contains(secret), "the description text must never reach the wire")
        // No description field exists anywhere in the contract; a stray one
        // would be rejected by the relay and is a privacy violation here.
        XCTAssertFalse(body.contains("\"description\""))
        // The length is the whole signal, and it is present.
        XCTAssertTrue(body.contains("\"length\":\(secret.count)"))
    }

    // MARK: - The contract cannot drift

    func testEventNamesMatchTheContractExactly() throws {
        let json = try XCTUnwrap(
            JSONSerialization.jsonObject(with: try ContractSource.dataFile("relay-events.json")) as? [String: Any]
        )
        let contractEvents = try XCTUnwrap(json["events"] as? [String: Any])
        XCTAssertEqual(Set(contractEvents.keys), Set(TelemetryEvent.allNames))
    }

    func testEventFieldsMatchTheContractExactly() throws {
        let json = try XCTUnwrap(
            JSONSerialization.jsonObject(with: try ContractSource.dataFile("relay-events.json")) as? [String: Any]
        )
        let contractEvents = try XCTUnwrap(json["events"] as? [String: Any])

        for name in TelemetryEvent.allNames {
            let contract = try XCTUnwrap(contractEvents[name] as? [String: Any], "contract names \(name)")
            let contractFields = Set(try XCTUnwrap(contract["fields"] as? [String: Any]).keys)
            let clientFields = Set(sampleEvent(forName: name).fields.keys)
            XCTAssertEqual(clientFields, contractFields, "field set for \(name) drifted from the contract")
        }
    }

    /// One representative event per name, used to pin the field sets. A new
    /// field on a case is a compile error here, which is the point: the drift
    /// is caught at build time, not by the relay's 400.
    private func sampleEvent(forName name: String) -> TelemetryEvent {
        switch name {
        case "app-opened": return .appOpened
        case "camera-permission": return .cameraPermission(granted: true)
        case "location-permission": return .locationPermission(granted: true)
        case "photo-captured": return .photoCaptured(elapsedMs: 0, bytes: 0, mime: "image/jpeg")
        case "location-resolved": return .locationResolved(elapsedMs: 0, source: .device, accuracyM: 0)
        case "location-failed": return .locationFailed(elapsedMs: 0, reason: .timeout)
        case "category-chosen": return .categoryChosen(elapsedMs: 0, slug: "ruslafotur")
        case "description-length": return .descriptionLength(length: 0)
        case "send-started": return .sendStarted
        case "send-result": return .sendResult(elapsedMs: 0, status: 200, ok: true)
        case "send-failed": return .sendFailed(elapsedMs: 0, reason: .connection)
        case "screen-left": return .screenLeft(screen: .camera, completed: true)
        default: XCTFail("unknown event name \(name)"); return .appOpened
        }
    }
}

// MARK: - The enum values themselves

/// The bug this file exists to prevent recurring.
///
/// The pinning tests above check event NAMES and FIELD KEYS. They passed while
/// `LocationFailure.noExif` went out as "noExif", because Swift's implicit raw
/// value is the case name and the contract says "no-exif". The relay 400s the
/// whole batch on an unknown enum value, and telemetry swallows its failures by
/// design, so the only symptom was that a class of event never appeared.
///
/// Kotlin cannot make this mistake: it spells every wire value explicitly. Swift
/// can, so Swift gets a test.
extension TelemetryTest {

    /// Every fixed-enum value the client can emit must be a value the contract
    /// names, and every value the contract names must be one the client can
    /// emit. Both directions: an unused contract value is dead vocabulary, and
    /// an unknown client value is a 400.
    func testEveryEnumValueMatchesTheContract() throws {
        let contract = try XCTUnwrap(
            JSONSerialization.jsonObject(with: try ContractSource.dataFile("relay-events.json"))
                as? [String: Any]
        )
        let events = try XCTUnwrap(contract["events"] as? [String: Any])

        func contractValues(event: String, field: String) throws -> Set<String> {
            let spec = try XCTUnwrap(events[event] as? [String: Any], "no event \(event)")
            let fields = try XCTUnwrap(spec["fields"] as? [String: Any])
            return Set(try XCTUnwrap(fields[field] as? [String], "\(event).\(field) is not an enum"))
        }

        let client: [(event: String, field: String, values: Set<String>)] = [
            (
                "location-resolved", "source",
                Set([TelemetryEvent.LocationSource.device, .exif].map(\.rawValue))
            ),
            (
                "location-failed", "reason",
                Set([
                    TelemetryEvent.LocationFailure.permission, .timeout, .unavailable, .noExif,
                ].map(\.rawValue))
            ),
            (
                "send-failed", "reason",
                Set([
                    TelemetryEvent.SendFailure.connection, .timeout, .encoding, .other,
                ].map(\.rawValue))
            ),
            (
                "screen-left", "screen",
                Set([TelemetryEvent.Screen.camera, .details, .confirm].map(\.rawValue))
            ),
        ]

        for entry in client {
            XCTAssertEqual(
                entry.values,
                try contractValues(event: entry.event, field: entry.field),
                "\(entry.event).\(entry.field) disagrees with data/relay-events.json"
            )
        }
    }

    /// The mime normaliser must land on a value the contract names, whatever it
    /// is handed. An iPhone shooting HEIC is the case that matters (#28).
    func testNormalizedMimeAlwaysLandsOnAContractValue() throws {
        let contract = try XCTUnwrap(
            JSONSerialization.jsonObject(with: try ContractSource.dataFile("relay-events.json"))
                as? [String: Any]
        )
        let events = try XCTUnwrap(contract["events"] as? [String: Any])
        let spec = try XCTUnwrap(events["photo-captured"] as? [String: Any])
        let fields = try XCTUnwrap(spec["fields"] as? [String: Any])
        let allowed = Set(try XCTUnwrap(fields["mime"] as? [String]))

        for raw in [
            "image/jpeg", "image/png", "image/gif", "image/heic",
            "image/heif", "application/pdf", "", "IMAGE/JPEG",
        ] {
            XCTAssertTrue(
                allowed.contains(Telemetry.normalizedMime(raw)),
                "normalizedMime(\(raw)) produced a value the relay would refuse"
            )
        }
    }
}
